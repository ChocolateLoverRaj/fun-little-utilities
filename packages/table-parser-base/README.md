# Table Parser Base

Utilities for table parser packages

## Usage

### Convert an `ArrayTable` to an `ObjectTable`

```javascript
const { createObjectTable } = await import('table-parser-base')

const arrayTable = {
  headers: ['id', 'name', 'email'],
  rows: [
    [1, 'John Doe', 'john-doe@gmail.com'],
    [2, 'Peter Smith', 'petersmith22@outlook.com'],
    [3, 'Julia Jones', 'jjones778@gmail.com']
  ]
}

const objectTable = createObjectTable(arrayTable)
for await (const item of objectTable) {
  console.log(item)
}
```

**Output:**

```javascript
{ id: 1, name: 'John Doe', email: 'john-doe@gmail.com' }
{ id: 2, name: 'Peter Smith', email: 'petersmith22@outlook.com' }
{ id: 3, name: 'Julia Jones', email: 'jjones778@gmail.com' }
```

### Convert an `ObjectTable` to an `ArrayTable`

```javascript
const { createArrayTable } = await import('table-parser-base')

const objectTable = [
  { id: 1, name: 'John Doe', email: 'john-doe@gmail.com' },
  { id: 2, name: 'Peter Smith', email: 'petersmith22@outlook.com' },
  { id: 3, name: 'Julia Jones', email: 'jjones778@gmail.com' }
]

const arrayTable = await createArrayTable(objectTable)
console.log(arrayTable)
```

**Output:**

```javascript
{
  headers: [ 'id', 'name', 'email' ],
  rows: [
    [ 1, 'John Doe', 'john-doe@gmail.com' ],
    [ 2, 'Peter Smith', 'petersmith22@outlook.com' ],
    [ 3, 'Julia Jones', 'jjones778@gmail.com' ]
  ]
}
```

## License

[MIT](https://git.io/fxKXN) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
