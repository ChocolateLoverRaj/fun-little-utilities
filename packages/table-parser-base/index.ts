import { zipAll } from 'iter-tools'
import AdvMapInit from 'advanced-map-initialized'

type MaybeAsyncIterable<X> = Iterable<X> | AsyncIterable<X>

export abstract class ArrayTable<Title extends string, Value> {
  public abstract readonly headers: readonly Title[]
  public abstract readonly rows: MaybeAsyncIterable<readonly Value[]>
}

export type ListItem<Key extends string, Value> = {
  readonly [_ in Key | symbol]: Value | undefined
}

class UnknownColumns extends AdvMapInit<number, symbol> {
  constructor () {
    super(Map, index => Symbol('Unknown Column #' + index))
  }
}

const UNKNOWN_COLUMNS = new UnknownColumns()
export const unknownColumn = (index: number) => UNKNOWN_COLUMNS.get(index)

export class ObjectTable<Title extends string, Value> implements AsyncIterable<ListItem<Title, Value>> {
  constructor (
    private readonly cells: ArrayTable<Title, Value>
  ) {}

  public async * [Symbol.asyncIterator] () {
    const { headers, rows } = this.cells

    for await (const row of rows) {
      const item: ListItem<Title, Value> = {} as any
      let index = 0

      for (const [key, value] of zipAll<any>(headers, row)) {
        (item as any)[key ? key : unknownColumn(index)] = value
        index += 1
      }

      yield item
    }
  }
}
