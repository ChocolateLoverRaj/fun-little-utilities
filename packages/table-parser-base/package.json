{
  "name": "table-parser-base",
  "version": "0.0.0",
  "description": "Utilities for table parser packages",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "sideEffects": false,
  "dependencies": {
    "iter-tools": "^6.2.6",
    "advanced-map-initialized": "^0.0.4",
    "tslib": "^1.10.0",
    "@types/node": "^13.5.2"
  }
}
