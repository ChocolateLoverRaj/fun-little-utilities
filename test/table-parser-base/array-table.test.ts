import { ObjectTable, ListItem, createArrayTable } from 'table-parser-base'

class Items<Item extends ListItem<string, any>> implements ObjectTable<string, any> {
  constructor(
    public readonly items: readonly Item[],
  ) {}

  public async *[Symbol.asyncIterator]() {
    yield* this.items
  }
}

it('no holes', async () => {
  const items = new Items([
    { id: 1, name: 'John Doe', email: 'john-doe@gmail.com' },
    { id: 2, name: 'Peter Smith', email: 'petersmith22@outlook.com' },
    { id: 3, name: 'Julia Jones', email: 'jjones778@gmail.com' },
  ])

  const table = await createArrayTable(items)
  expect(table).toMatchSnapshot()
})

it('with holes', async () => {
  const items = new Items([
    { id: 1, name: 'John Doe', age: 32 },
    { id: 2, name: 'Peter Smith', sex: 'male', alive: true },
    { id: 3, name: 'Julia Jones', email: 'jjones778@gmail.com' },
  ])

  const table = await createArrayTable(items as any)
  expect(table).toMatchSnapshot()
})
