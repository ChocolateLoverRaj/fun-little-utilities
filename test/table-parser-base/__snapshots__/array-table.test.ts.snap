// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no holes 1`] = `
Object {
  "headers": Array [
    "id",
    "name",
    "email",
  ],
  "rows": Array [
    Array [
      1,
      "John Doe",
      "john-doe@gmail.com",
    ],
    Array [
      2,
      "Peter Smith",
      "petersmith22@outlook.com",
    ],
    Array [
      3,
      "Julia Jones",
      "jjones778@gmail.com",
    ],
  ],
}
`;

exports[`with holes 1`] = `
Object {
  "headers": Array [
    "id",
    "name",
    "age",
    "sex",
    "alive",
    "email",
  ],
  "rows": Array [
    Array [
      1,
      "John Doe",
      32,
    ],
    Array [
      2,
      "Peter Smith",
      undefined,
      "male",
      true,
    ],
    Array [
      3,
      "Julia Jones",
      undefined,
      undefined,
      undefined,
      "jjones778@gmail.com",
    ],
  ],
}
`;
